### ISSUE TITLE
[2016.11.1] Salt-cloud EC2 failing to launch new instances

### ISSUE THREAD

**Issue by tyhunt99** [saltstack/salt#38351]
### Description of Issue/Question
Not specifying a   block_device_mappings section causes an error and instances won't launch. It produces the following error:
```
Traceback (most recent call last):
  File "/usr/local/bin/salt-cloud", line 11, in <module>
    sys.exit(salt_cloud())
  File "/usr/local/lib/python2.7/dist-packages/salt/scripts.py", line 439, in salt_cloud
    client.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/cli.py", line 283, in run
    self.config.get('names')
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1458, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1288, in create
    output = self.clouds[func](vm_)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 2516, in create
    data, vm_ = request_instance(vm_, location)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 1896, in request_instance
    if 'Ebs.VolumeType' not in ex_blockdevicemappings[dev_index]:
TypeError: 'NoneType' object has no attribute '__getitem__'
Traceback (most recent call last):
  File "/usr/local/bin/salt-cloud", line 11, in <module>
    sys.exit(salt_cloud())
  File "/usr/local/lib/python2.7/dist-packages/salt/scripts.py", line 439, in salt_cloud
    client.run()
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/cli.py", line 283, in run
    self.config.get('names')
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1458, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/__init__.py", line 1288, in create
    output = self.clouds[func](vm_)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 2516, in create
    data, vm_ = request_instance(vm_, location)
  File "/usr/local/lib/python2.7/dist-packages/salt/cloud/clouds/ec2.py", line 1896, in request_instance
    if 'Ebs.VolumeType' not in ex_blockdevicemappings[dev_index]:
TypeError: 'NoneType' object has no attribute '__getitem__'
```
\* I also couldn't get salt cloud to honor my master config when it comes to the debug level not sure why so I removed a catch temporarily to get this stack trace.

I manged to track down the issue to a specific commit: #33115. This changed this line of code from this
```
# version 2016.3.4
if rd_type is not None:
```
to 
```
# version 2016.11.1
if 'Ebs.VolumeType' not in ex_blockdevicemappings[dev_index]:
```

Reverting that line fixes at least this issue. But I also believe there is an index out of range error caused when you do specify the block_device_mapping. I will investigate further.

### Setup
Here is my ec2 cloud profile:
```
data_loader_large:
  provider: petrode-ec2
  image: ami-fce3c696
  size: m4.xlarge
  ssh_username: ubuntu
  ssh_interface: private_ips

  ebs_optimized: True

  subnetid: subnet-1caa6e37

  tag: {'Managed': 'Salt', 'Role': 'Dataload', 'Environment': 'Production' }

  grains:
    role: dataload

  securitygroupid:
    - sg-e18a7299
    - sg-cef270b7
    - sg-bfedeec6
    - sg-f1ec6e88

  volumes:
    - { size: 150, device: /dev/xvdc, type: gp2 }
    - { size: 100, device: /dev/xvdd, type: gp2 }
    - { size: 100, device: /dev/xvde, type: gp2 }
    - { size: 100, device: /dev/xvdf, type: gp2 }
    - { size: 100, device: /dev/xvdg, type: gp2 }
    - { size: 100, device: /dev/xvdh, type: gp2 }
    - { size: 100, device: /dev/xvdi, type: gp2 }
    - { size: 100, device: /dev/xvdj, type: gp2 }
    - { size: 100, device: /dev/xvdk, type: gp2 }

  # force grains to sync after install
  sync_after_install: grain
```

### Versions Report
(Provided by running `salt --versions-report`. Please also mention any differences in master/minion versions.)
```
Salt Version:
           Salt: 2016.11.1
 
Dependency Versions:
           cffi: 1.5.0
       cherrypy: 3.2.2
       dateutil: 2.6.0
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.23.4
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: 1.2.3
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: 0.23.1
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: 0.3.6
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.1.6
 
System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-100-generic
         system: Linux
        version: Ubuntu 14.04 trusty
```

**Comment by Ch3LL** [comment]
@tyhunt99 can you give this fix a try: #37967 ?

**Comment by tyhunt99** [comment]
@Ch3LL yup that appears to fix this issue
